rabbitmq/retryable_consumer.go

```diff
@@ -0,0 +1,123 @@
+// Copyright 2019 SumUp Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rabbitmq
+
+import (
+	"context"
+	"time"
+
+	"github.com/palantir/stacktrace"
+	"go.uber.org/zap"
+
+	"github.com/sumup-oss/go-pkgs/backoff"
+	"github.com/sumup-oss/go-pkgs/logger"
+)
+
+type RetryableConsumer struct {
+	config        RetryableConsumerConfig
+	logger        logger.StructuredLogger
+	metric        Metric
+	handler       Handler
+	clientFactory func(ctx context.Context, config *ClientConfig) (*RabbitMQClient, error)
+}
+
+type RetryableConsumerConfig struct {
+	MaxRetryAttempts int
+	// healthCheckFactor is a number representing how much N multiplied by backoffConfig.Max time is needed
+	// for a block of code to run w/o returning an error, to consider it healthy.
+	// E.g backConfig.Max = 1min, healthCheckFactor = 2, means that code needs to run 2min at least to be healthy
+	// and retried again starting from backoffConfig.Base the next time it has an error.
+	HealthCheckFactor  int
+	BackoffConfig      *backoff.Config
+	ConsumerConfig     ConsumerConfig
+	RabbitClientConfig *ClientConfig
+}
+
+func NewRetryableConsumer(
+	newClientFactory func(ctx context.Context, config *ClientConfig) (*RabbitMQClient, error),
```

Comment:
If we want to improve the testability, I believe we should return an interface rather than a concrete type as is the case with *RabbitMQClient in the factory method

Reply:
Didn't we discuss to use factory func instead of interface? Otherwise the whole parameter can be interface that complies with RabbitMQClient. Anyway I think it is better to be factory method since I have a good usage for it inside one service so I am able to pass my own func that runs the RabbitMQ setup inside and returns the client.

Reply:
I meant for the factory method's signature to return an interface for the rabbit client rather than the concrete type of the client we're using.

Reply:
I may be wrong here but I also think that returning an interface is not the best practice.

Reply:
Because otherwise if you try to replace the factory method with one which would return a mock, then it wouldn't work since you can't mock a concrete type

Reply:
Yeah I got it. You think of something like this `newClientFactory func(ctx context.Context, config *ClientConfig) (RabbitClientInterface, error)`. I am not sure it is this we want though.

Reply:
If we don't do this, we can't really test the library I'm afraid. At least I'm not aware of how to do it

Reply:
If we do this this should be inside the rabbit client package IMO also.

Reply:
And also that the other main producer and consumer should be change to use interface instead of client.

Reply:
If we all agree to change those parts I can make it real quick and add some tests around this. Not a big change for the producers and consumers IMO. I agree code will be more testable that way.

Reply:
Added interface and refactored the other producer and consumers. Others should approve also since this change affects a bit the main producer and consumer.